src/pe/delay_import.rs

```diff
@@ -0,0 +1,529 @@
+    pub bound_table_rva: u32,
+    /// RVA of the unload import address table (optional; used during DLL unloading).
+    pub unload_table_rva: u32,
+    /// Timestamp of the bound DLL; zero if not bound.
+    pub time_stamp: u32,
+}
+
+impl<'a> DelayImportDescriptor {
+    /// Returns whether the RVA is based (RvaBased field).
+    pub fn rva_based(&self) -> bool {
+        self.attributes & 0x1 != 0
+    }
+
+    /// Returns the ReservedAttributes field (31 bits).
+    pub fn reserved_attributes(&self) -> u32 {
+        self.attributes >> 1
+    }
+
+    /// Whether the entire fields are set to zero
+    pub fn is_null(&self) -> bool {
+        self.attributes == 0
+            && self.name_rva == 0
+            && self.handle_rva == 0
+            && self.address_table_rva == 0
+            && self.name_table_rva == 0
+            && self.bound_table_rva == 0
+            && self.unload_table_rva == 0
+            && self.time_stamp == 0
+    }
+
+    /// Whether the entry is _possibly_ valid.
+    ///
+    /// Both [`Self::name_rva`] and [`Self::address_table_rva`] must be non-zero
+    pub fn is_possibly_valid(&self) -> bool {
+        self.name_rva != 0 && self.address_table_rva != 0
+    }
+}
+
+/// Represents a PE delay import directory data.
+#[derive(Debug)]
+pub struct DelayImportData<'a> {
+    ctx: PeCtx<'a>,
+    bytes: &'a [u8],
+    /// Delay imported entries.
+    pub imports: Vec<DelayImportEntry<'a>>,
+}
+
+impl<'a> DelayImportData<'a> {
+    pub fn parse(
+        bytes: &'a [u8],
+        dd: data_directories::DataDirectory,
+        sections: &[section_table::SectionTable],
+        file_alignment: u32,
+        is_64: bool,
+    ) -> error::Result<DelayImportData<'a>> {
+        Self::parse_with_opts(
+            bytes,
+            dd,
+            sections,
+            file_alignment,
+            &options::ParseOptions::default(),
+            is_64,
+        )
+    }
+
+    pub fn parse_with_opts(
+        bytes: &'a [u8],
+        dd: data_directories::DataDirectory,
+        sections: &[section_table::SectionTable],
+        file_alignment: u32,
+        opts: &options::ParseOptions,
+        is_64: bool,
+    ) -> error::Result<DelayImportData<'a>> {
+        let offset =
+            utils::find_offset(dd.virtual_address as usize, sections, file_alignment, opts)
+                .ok_or_else(|| {
+                    error::Error::Malformed(format!(
+                        "cannot map delay import table {:#x}",
+                        dd.virtual_address
+                    ))
+                })?;
+        if offset + dd.size as usize > bytes.len() {
+            return Err(error::Error::Malformed(format!(
+                "delay import offset {:#x} and size {:#x} exceeds the bounds of the bytes size {:#x}",
+                offset,
+                dd.size,
+                bytes.len()
+            )));
+        }
+        let directory_bytes = &bytes[offset..offset + dd.size as usize];
```

Comment:
rewrite this to `pread::<&[u8]>(offset, dd.size as usize).map_err` and can remove check above and replace error with the `map_err`